txn: add document for read-consistency read tso optimization (#32806)
pingcap/tidb: docs/design/2022-03-03-rc-read-tso-optimization.md

# Read-Consistency Read With Timestamp Check

- Author(s): cfzjywxk
- Last updated: March. 3, 2022
- Discussion at:

## Motivation

> For the read-consistency isolation level, each read request in a single transaction will need to fetch a new `ts` to read the latest committed data.

Reviewer: I perfer `read-committed` to `read-consistency`.

If the workload is a read-heavy one whose read QPS is significantly higher than writes or there're few read-write conflicts, fetching ts each time will increase the query lantecy.

The new ts itself is used to ensure the most recent data will be returned, if the data version does not change frequently then it's unnecessary to fetch a new timestamp every time.
The ts fetching could be processed in an optimistic way that ts fetching happens only when a more recent data version is encoutered, this saves the cost of many unncessary ts fetching.

## Detailed Design

> Introduce another system variable `tidb_rc_read_check_ts` to enable the "lazy ts fetch" read mode. It will take effect for all the in-transaction select statements when the transaction mode is `pessimistic` and the isolation level is `read-consistency`.

Reviewer: > the isolation level is `read-consistency`

The behavior is `read-consistency`, but the isolation level is `read-committed`.

1. If the query is executed first time, do not fetch a new `for_update_ts` and just use the last valid one(`start_ts` for the first time).
2. Do build the plan and executor like before.
3. If the execution is successful, a `resultSet` will be returned.
4. The connection layer will drive the response procesure and call `Next` function using the returned `resultSet`.
5. The read executor will try to fetch data to fill in a data `chunk`. The bottommost executor could be a `pointGet`, `batchPointGet` or a coprocessor task. What's different is a `RcReadCheckTS` flag will be set on all these read requests.
6. The read executor in the storage layer will check the read results, if a more recent version does exist then a `WriteConflict` error is reported.
7. For data write record, do check if it's newer than the current `read_ts`.
8. For lock record, return `ErrKeyIsLocked` even though the `lock.ts` is greater than the `read_ts` as the `read_ts` could be a stale one.
9. If no error is returned then the query is finished. Otherwise if there's no `chunk` responsed to the client yet, retry the whole query fetching a new global ts to do the read like normal.

## Compatibility

> The default behaviours of the `read-consistency` isolation level will not change. One thing different is that if the user client uses `COM_STMT_FETCH` like utility to read data from `TiDB`,

Reviewer: ditto.

# Read-Consistency Read With Tso Check

- Author(s): cfzjywxk
- Last updated: March. 3, 2022
- Discussion at:

## Motivation

> For the read-consistency isolation level read requests in a single transaction, each will need to fetch a new `tso` to read the latest committed data.

Reviewer 2: `tso` refers to the allocator instead of a timestamp

> For the read-consistency isolation level read requests in a single transaction, each will need to fetch a new `tso` to read the latest committed data.

Reviewer 3: read-consistency? is it read-committed?

Reviewer 2: It actually refers to the term "statement-level read consistency" from Oracle. Yes, seems better to clarify it's the semantics of "read committed" in TiDB.

> For the read-consistency isolation level read requests in a single transaction, each will need to fetch a new `ts` to read the latest committed data.

Reviewer 3: read-consistency? is it a read-committed isolation?

> For the read-consistency isolation level read requests in a single transaction, each will need to fetch a new `ts` to read the latest committed data.

Reviewer 4: ```suggestion
For the read-consistency isolation level, each read request in a single transaction will need to fetch a new `ts` to read the latest committed data.
```

> If the workload is a read-heavy one or the read `qps` is large, fetching tso each time will increase the query lantecy.

Reviewer 5: Why the words such as "tso" and "qps" are in inline code format (in backquotes)?

> If the workload is a read-heavy one or the read qps is large, fetching ts each time will increase the query lantecy.

Reviewer 4: Does "the read qps is large" means "the read QPS is significantly higher than writes"?

> The new tso itself is used to ensure the most recent data will be returned, if the data version dose not change frequently then it's unnecessary to fetch tso every time.

Reviewer: ```suggestion
The new tso itself is used to ensure the most recent data will be returned, if the data version does not change frequently then it's unnecessary to fetch tso every time.
```

> That is the `rc-read` could be processed in a optimistic way, the tso could be updated only when a new version data is met， then the tso cost will be saved a lot for this case.

Reviewer: ```suggestion
That is the `rc-read` could be processed in an optimistic way, the tso could be updated only when a new version data is met， then the tso cost will be saved a lot for this case.
```

> That is the `rc-read` could be processed in a optimistic way, the tso could be updated only when a new version data is met， then the tso cost will be saved a lot for this case.

Reviewer: ```suggestion
That is the `rc-read` could be processed in a optimistic way, the tso could be updated only when a new version data is met, then the tso cost will be saved a lot for this case.
```

## Detailed Design

> Introduce another system variable `tidb_rc_read_check_ts` to enable the "lazy ts update" read mode. It will take effect for all the in-transaction select stataments and the

Reviewer 4: ```suggestion
Introduce another system variable `tidb_rc_read_check_ts` to enable the "lazy ts update" read mode. It will take effect for all the in-transaction select statements when the
```

transaction isolation level is `read-consistency`.

The execution flow is like this:

1. If the query is executed first time, do not fetch a newer `for_update_ts` and just use the last valid one(`start_ts` for the first time).
2. Do build the plan and executor like before.
3. If the execution is successful, a `resultSet` will be returned.
4. The connection layer will drive the response and call `Next` function using the `resultSet`.
5. The read executor will try to fetch data to fill in a `chunk`. The bottommost executor could be a `pointGet`, `batchPointGet` or coprocessor task. What's different is a `RcReadCheckTS` flag will be set on the read requests.

> 6. The read executor in storage layer will check the read results, if more recent version does exist then a `WriteConflict` error will be returned.

Author: It could be regarded as that a newer write has conflicted with the current read, so I just reused the error code.

> 6. The read executor in storage layer will check the read results, if more recent version does exist then a `WriteConflict` error will be returned.

Reviewer 4: ```suggestion
6. The read executor in the storage layer will check the read results, if a more recent version does exist then a `WriteConflict` error will be returned.
```

7. For data write record, do check if there's new version compared with current `read_ts`.
8. For lock record, return `ErrKeyIsLocked` though the `lock.ts` is greater than the `read_ts` as the `read_ts` could be a stale one.

> 9. If no error is returned the query is finished. Otherwise if there's no `chunk` responsed to client yet, do retry the whole query and this time a new `for_update_ts` will be fetched.

Reviewer: So we don't fetch a ts in the background at the beginning?

Author: Yes, in the last requirement disccussion the async fetch is required to be avoided.

> 9. If no error is returned the query is finished. Otherwise if there's no `chunk` responsed to client yet, do retry the whole query and this time a new `for_update_ts` will be fetched.

Reviewer 2: What error will be returned to the client if some chunks have been responded to the client?

Author: Write conflit error will be returned to the client.

Reviewer 6: Is it acceptable to users?

Author: In some situations the early returned result could be alredy used for example when the cursor is used by the mysql client, reporting error could not help with this. It's not recomanded to use this feature in such user case, or the `init_chunk_size` needs to be increased so the results are always batched in tidb-server before finishing all the checks.

> 9. If no error is returned the query is finished. Otherwise if there's no `chunk` responsed to client yet, do retry the whole query and this time a new `for_update_ts` will be fetched.

Reviewer 6: I think since tidb has to read it twice, why not read it the first time with maxint64, and then, calculate the tso brought back in the read response as a max+1 and read it again. This should avoid getting tso in RC?

Reviewer 2: The timestamp in the read response is not certainly new enough. It is possible that there is a write record with larger commit TS in the third shard. And if the error is caused by a lock, we actually cannot know its commit TS.

Getting a new ts from TSO works simple and correct.

The only thing I might do different here is to get ts parallelly with the first read. I believe it won't increase much load thanks to the tso batching.

## Compatibility

> The default behaviours of the `read-consistency` isolation level will not change. One thing differnt is that if the user client uses `COM_STMT_FETCH` like utility to read data from `TiDB`,

Reviewer: ```suggestion
The default behaviours of the `read-consistency` isolation level will not change. One thing different is that if the user client uses `COM_STMT_FETCH` like utility to read data from `TiDB`,
```

> there could be problem if the returned first chunk result is already used by the client but an error is reported processing next result chunk.

Reviewer: I think we should elaborate on it.

Author: Yes it's needed.
